api/v1alpha3/conversion.go

```diff
@@ -96,13 +98,38 @@ func (dst *MachineSetList) ConvertFrom(srcRaw conversion.Hub) error {
 func (src *MachineDeployment) ConvertTo(dstRaw conversion.Hub) error {
 	dst := dstRaw.(*v1alpha4.MachineDeployment)
 
-	return  Convert_v1alpha3_MachineDeployment_To_v1alpha4_MachineDeployment(src, dst, nil)
+	if err := Convert_v1alpha3_MachineDeployment_To_v1alpha4_MachineDeployment(src, dst, nil); err != nil {
+		return err
+	}
+
+	// Manually restore data.
+	restored := &v1alpha4.MachineDeployment{}
+	if ok, err := utilconversion.UnmarshalData(src, restored); err != nil || !ok {
+		return err
+	}
+
+	if restored.Spec.Strategy != nil && restored.Spec.Strategy.RollingUpdate != nil {
+		if dst.Spec.Strategy != nil && dst.Spec.Strategy.RollingUpdate != nil {
```

Comment:
If Strategy, or Strategy.RollingUpdate is nil in the converted function, but we have something to restore, we should make sure to set it

Reply:
IIRC, doing that was causing the new `spoke-hub-spoke` test to fail because if the initial strategy was nil, it expects the strategy to be nil after the whole roundtrip of conversions.

Reply:
I think the use case it fails is:
hub with strategy!=nil -> spoke1 created from hub (with strategy) -> spoke1 modified to strategy nil -> spoke1 to hub -> hub to spoke2 -> spoke2 expects strategy to be nil even though hub has strategy.

Reply:
Getting a bit confused 😄 , if `DeletePolicy` is set in the `restored` object, we should make sure to always set it in `dst`, otherwise we'd have data loss.

If by doing that the `spoke-hub-spoke` test fails, we'd have to audit the test 🤔 because that shouldn't happen.

Was thinking something along this: (haven't tested this yet)

```go
	if restored.Spec.Strategy != nil && restored.Spec.Strategy.RollingUpdate != nil && dst.Spec.Strategy.RollingUpdate.DeletePolicy != nil {
		if dst.Spec.Strategy == nil {
			dst.Spec.Strategy = &MachineDeploymentStrategy{}
		}
		if dst.Spec.Strategy.RollingUpdate == nil {
			dst.Spec.Strategy.RollingUpdate = &MachineRollingUpdateDeployment{}
		}
		dst.Spec.Strategy.RollingUpdate.DeletePolicy = restored.Spec.Strategy.RollingUpdate.DeletePolicy
```

Reply:
I know, it's totally confusing :) Sorry, I couldn't explain it better. 
The code above fails because there is a test case where restored(hub) has `strategy != nil`, but `spoke1.strategy = nil`. So, spoke2 is expected to have `strategy == nil`. Same applies for anything inside `MachineDeploymentStrategy` Failure logs ->
```
&v1alpha3.MachineDeployment{
          	TypeMeta:   {},
          	ObjectMeta: {Name: "epȢ妅璴鯹pj暕ǅ´澿", GenerateName: "c璐Ñ浥Þ(T褊7T#ȦK逾蚲", Namespace: "ÀyȎǾ", SelfLink: "Ǜť[镏!HɼÅ唱E聢Ȑ鉁{", ...},
          	Spec: v1alpha3.MachineDeploymentSpec{
          		... // 2 identical fields
          		Selector: {MatchExpressions: {{Key: "9yi93xn4sf8---g837wgt-p-rfwn12-7i8-lm035-5-s-2r-j--9-kp6vjs/23_c"..., Operator: "In", Values: {"5-hqIJ._-_V-_z.0..0DE", "UW-e_C.vb.-N__0ct-oB_6..y-Zg-C"}}}},
          		Template: {ObjectMeta: {Name: "磕Ʃ帓倂ŵ@6鹒婰", GenerateName: "厧", Namespace: "0峥ȶlɒq抹ʞ嫀渄L塪>ē@毝", Annotations: {"ƪŵɽ19": "ɲņǮ嫙櫄E"}, ...}, Spec: {ClusterName: "鷮n秂ŋc抙滰W;Ȯ翤猫Ť鸊ĤÒĦ", Bootstrap: {ConfigRef: &{Kind: "Ě/žxÓ", Namespace: "Fǩæ疎06ʤɯ磼ǀNƓ謚ʪ懨vfv.", Name: "ɒɶ鍱勬%!唲(MISSING)imɊ烮yM8Â肙,", UID: "jƒ", ...}, Data: &"K滏歊镎RǓt腓誥è柜oƲījY莼"}, InfrastructureRef: {Kind: "ɣ4", Namespace: "5ƪ柄", Name: "'b慾.脣ŝ彿_mºȸ淥ļ", UID: "Ǻ摍啨W¿egC疟4迅ôRe*p慉N", ...}, Version: &"|M", ...}},
          		Strategy: &v1alpha3.MachineDeploymentStrategy{
          			Type:          "{`Ǒ",
        - 			RollingUpdate: nil,
        + 			RollingUpdate: &v1alpha3.MachineRollingUpdateDeployment{},
          		},
          		MinReadySeconds:      &615253399,
          		RevisionHistoryLimit: &-1212597792,
          		... // 2 identical fields
          	},
          	Status: {ObservedGeneration: -2506154037848954697, Selector: "饷塅<oVp$鸬Ĳ鎟ûB", Replicas: 1626262098, UpdatedReplicas: -853823741, ...},
          }
        
        Expected
            <bool>: false
        to be true
```
As you can see `RollingUpdate` is expected to be nil because spoke1 was fuzzed to have rollingupdate as nil. Does that make sense?

Reply:
The test seems it's causing us to lose data in this instance, I think we might have to revisit it

cc @ncdc

Reply:
In the spoke-hub-spoke test, the `spokeFirstGet` object contains fuzzed annotation data (i.e. the hub values) plus different fuzzed spec values. The purpose of this test is to ensure that all `spokeFirstGet.spec` data is preserved and not overwritten by the fuzzed annotation data; i.e., it's identical to `spokeSecondGet.spec`.

In this specific case, if `spokeFirstGet.spec.strategy` is nil, but the fuzzed annotation data has a non-nil `strategy`, the expected behavior should be for `spokeFirstGet.spec.strategy` to also be nil. The only field we should ever be restoring in the annotation data restoration flow is `spec.strategy.rollingUpdate.deletePolicy`. Right?

Reply:
@vincepri with #3963 merged, I've updated the conversion logic to always set `DeletePolicy` in `dst` object if it is present in the `restored` object, PTAL.